lib/devicetrust/native/device_windows.go

```diff
@@ -511,10 +535,170 @@ func solveTPMEnrollChallenge(
+// elevated privileges in order to invoke the TPM 2.0 ActivateCredential
+// command.
+func activateCredentialInElevatedChild(
+	ctx context.Context,
+	encryptedCredential attest.EncryptedCredential,
+	credActivationPath string,
+) ([]byte, error) {
+	exe, err := os.Executable()
+	if err != nil {
+		return nil, trace.Wrap(err, "determining current executable path")
+	}
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, trace.Wrap(err, "determining current working directory")
+	}
+
+	// Clear up the results of any previous credential activation
+	if err := os.Remove(credActivationPath); err != nil {
+		err := trace.ConvertSystemError(err)
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err, "clearing previous credential activation results")
+		}
+	}
+
+	// Assemble the parameter list. We encoded any binary data in base64.
+	// The list is then converted to a space delimited string.
+	// These parameters cause `tsh` to invoke HandleTPMActivateCredential.
+	params := strings.Join([]string{
+		"-d", // enable debug mode so relevant logs appear in the new window
+		"device",
+		"tpm-activate-credential",
+		"--encrypted-credential",
+		base64.StdEncoding.EncodeToString(encryptedCredential.Credential),
+		"--encrypted-credential-secret",
+		base64.StdEncoding.EncodeToString(encryptedCredential.Secret),
+	}, " ")
+
+	// Create pointers to invoke the windows syscall with.
+	operation := "runas"
+	lpOperation, err := syscall.UTF16PtrFromString(operation)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting operation to ptr")
+	}
+	lpFile, err := syscall.UTF16PtrFromString(exe)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting exe to ptr")
+	}
+	lpParameters, err := syscall.UTF16PtrFromString(params)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting params to ptr")
+	}
+	lpDirectory, err := syscall.UTF16PtrFromString(cwd)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting cwd to ptr")
+	}
+
+	log.Debug("Starting elevated process.")
+	// https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecutew
+	err = windows.ShellExecute(
+		0, // hwnd
+		lpOperation,
+		lpFile,
+		lpParameters,
+		lpDirectory,
+		windows.SW_NORMAL,
+	)
+	// The windows.ShellExecute blocks until the user answers the UAC dialogue.
+	// If they reject the UAC dialogue, an error is returned:
+	// * syscall.Errno: The operation was canceled by the user.
+	if err != nil {
+		return nil, trace.Wrap(err, "invoking ShellExecute")
+	}
+
+	// Ensure we clean up the results of the execution once we are done with
+	// it.
+	defer func() {
+		if err := os.Remove(credActivationPath); err != nil {
+			log.WithError(err).Debug("Failed to clean up credential activation result")
+		}
+	}()
+
+	log.Debug("Waiting for credential activation solution from elevated process.")
+	// Wait for result from child process, polling every half second for up to
+	// 10 seconds.
+	for i := 0; i < 20; i++ {
```

Comment:
A more elegant approach could be to use [ShellExecuteEx](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecuteexw) to receive the created process handle as an out parameter and wait on it with `WaitForSingleObject`. Obviously you would want to return with an error if it times out, but beats polling for a created file.

Reply:
Do you know if `ShellExecuteEx` is exposed in the Go `windows` package? As far as I can tell all they expose at the moment is `ShellExecuteW` through `windows.ShellExecute` and unsure if there's a reason they've chosen to emit ShellExecuteEx

Reply:
A bit more cgo won't make much of difference at this stage, if you want to pursue it. Maybe as a follow up to this PR?

Reply:
Probably didn't need it, they add new functions all the time